File: frontend/src/game/upgrades/curses/shattered_bullet_3.ts
import { Upgrade, type UpgradeDef } from '../Upgrade'
import { UpgradeTargetID, UpgradeStatID, RarityID, UpgradeTypeID } from '../../data/ID'

export const ShatteredBullet3Def: UpgradeDef = {
  id: "shattered_bullet_3",
  name: "Shattered Bullet 3",
  description: "-5 bullet damage.",
  rarity: RarityID.Epic,
  upgradeType: UpgradeTypeID.StatModifier,
  cost: 0,
  targetClass: UpgradeTargetID.Bullet,
  fieldInTargetClass: UpgradeStatID.Damage,
  value: -5,
  isMultiplier: false,
  stackable: true,
  maxStacks: 99999,
  curse: true,
}

export class ShatteredBullet3 extends Upgrade {}

File: backend/app/core/upgrades/curses/shattered_bullet_3.py
from ..upgrade_implementation import UpgradeImplementation, UpgradeAppliers


class ShatteredBullet3(UpgradeImplementation):
    id = 'shattered_bullet_3'
    name = 'Shattered Bullet 3'
    description = '-5 bullet damage.'
    rarity = 'epic'
    type = 'stat_modifier'
    cost = 0

    def can_apply(self, state: dict) -> bool:
        return True

    def apply(self, appliers: UpgradeAppliers) -> None:
        value = -5
        appliers.modifiers.add_multiplier('bullet', 'damage', value)

    def validate(self, recorded_stats: dict) -> bool:
        return True
